### PR TITLE
Issue 130: Support multiple jupyter extension in Leo

### PR DESCRIPTION
**Pull request by vkumra-broad**
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 1]
```diff
@@ -315,5 +315,22 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
         }
       }
     }
+
+    "should install multiple user specified notebook extensions" in withWebDriver { implicit driver =>
+      val clusterName = ClusterName("user-jupyter-ext" + makeRandomId())
+      withNewCluster(billingProject, clusterName, ClusterRequest(Map(), None, None, None, Some(UserJupyterExtensionConfig(Map("translate"->testJupyterExtensionUri, "map"->"gmaps"),Map("jupyterlab"->"jupyterlab"), Map("pizza"->"pizzabutton"))))) { cluster =>
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: Minor syntax comment: you can specify just a single argument like:
```
ClusterRequest(billingProject, clusterName, userJupyterExtensionConfig = Some(extensionConfig))
```
Might also be good to define the extension config in a separate variable so it's easier to see what's being installed.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 2]
```diff
@@ -315,5 +315,22 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
...
+        withNewNotebook(cluster, Python3) { notebookPage =>
+
+          //Check if the mark up was translated correctly
+          val nbExt = notebookPage.executeCell("! jupyter nbextension list")
+          nbExt.get should include("jupyter-gmaps/extension  enabled")
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: Does the actual output have 2 spaces before `enabled`?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 3]
```diff
@@ -315,5 +315,22 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
...
+          nbExt.get should include("pizzabutton/index  enabled")
+          nbExt.get should include("my_ext/main  enabled")
+          val serverExt = notebookPage.executeCell("! jupyter serverextension list")
+          serverExt.get should include("pizzabutton  enabled")
+          serverExt.get should include("jupyterlab  enabled")
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: I definitely want to try out jupyterlab myself after reviewing this PR

File: src/main/resources/jupyter/init-actions.sh [hunk 4]
```diff
@@ -55,6 +55,9 @@ if [[ "${ROLE}" == 'Master' ]]; then
     JUPYTER_DOCKER_COMPOSE=$(jupyterDockerCompose)
     JUPYTER_PROXY_SITE_CONF=$(jupyterProxySiteConf)
     JUPYTER_EXTENSION_URI=$(jupyterExtensionUri)
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/resources/jupyter/init-actions.sh]: Could we make it if `jupyterExtensionUri` is specfied in the API then we just append it to the `JUPYTER_NB_EXTENSIONS` list? Then we have backwards compatibility in the API but don't have to maintain separate paths in the init script.

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -653,6 +653,9 @@ definitions:
       labels:
         type: object
         description: The labels to be placed on the cluster. Of type Map[String,String]
+      userJupyterExtensionConfig:
+        $ref: "#/definitions/UserJupyterExtensionConfig"
+        description: The Jupyter Extensions to be placed on the cluster.
```
Comment (rtitle) [inline src/main/resources/swagger/api-docs.yaml]: Maybe "Jupyter extensions to be installed in the notebook" sounds better?

File: src/main/resources/swagger/api-docs.yaml [hunk 6]
```diff
@@ -719,7 +722,19 @@ definitions:
...
           Optional, the number of preemptible workers. If unspecified, the default number is 0. Ignored if numberOfWorkers is 0.
           For more information, see https://cloud.google.com/compute/docs/instances/preemptible
-
+  UserJupyterExtensionConfig:
+    description: 'Specification of Jupyter Extensions to be installed on the cluster'
+    properties:
+      nbExtensions:
+        type: object
+        description: |
+          Optional, map of extension name and notebook
```
Comment (rtitle) [inline src/main/resources/swagger/api-docs.yaml]: I think some more explanatory text and examples would be useful here. Maybe also mention that both .tar.gz files and python packages are supported.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 7]
```diff
@@ -327,9 +327,27 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
       _ <- validateBucketObjectUri(userEmail, googleProject, clusterRequest.jupyterExtensionUri)
       _ <- validateBucketObjectUri(userEmail, googleProject, clusterRequest.jupyterUserScriptUri)
+
+      _ <- clusterRequest.userJupyterExtensionConfig match {
+        case Some(x) => Future.traverse(x.nbExtensions)(x => validateBucketObjectUri(userEmail, googleProject, parseGcsPath(x._2).toOption))
+        case None => Future.successful(())
+      }
+
+      _ <- clusterRequest.userJupyterExtensionConfig match {
+        case Some(x) => Future.traverse((x.serverExtensions).filterKeys(_.startsWith("gs://")))(x => validateBucketObjectUri(userEmail, googleProject, parseGcsPath(x._2).toOption))
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Is `.filterKeys` right? I thought the keys were names, and the validates were gs:// paths.

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: What if we did this:
```
_ <- clusterRequest.userJupyterExtensionConfig match {
  case Some(config) =>
    val extensionsToValidate = (config.nbExtensions.values ++ config.serverExtensions.values ++ config.combinedExtensions.values).filter(_.startsWith("gs://"))
    Future.traverse(extensionsToValidate)(x => validateBucketObjectUri(userEmail, googleProject, x))
   case None => Future.successful(())
}
```
and made `validateBucketObjectUri` take a `String` instead of a `Option[GcsPath]`. Then `validateBucketObjectUri` could do both `parseGcsPath()` _and_ `storageDAO.objectExists()` and throw an exception if either one fails.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 8]
```diff
@@ -437,9 +455,26 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
 
   /* Process the templated cluster init script and put all initialization files in the init bucket */
   private[service] def initializeBucketObjects(userEmail: WorkbenchEmail, googleProject: GoogleProject, clusterName: ClusterName, initBucketName: GcsBucketName, clusterRequest: ClusterRequest, serviceAccountKey: Option[ServiceAccountKey]): Future[Unit] = {
+
+
+    val jupyterNbExtensions = clusterRequest.userJupyterExtensionConfig match {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Am wondering if this logic should go in `ClusterInitValues.apply` instead.

Also could make a generic function:
```
def templateMap[A, B](map: Map[A, B]): String = map.values.mkString(" ")
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 9]
```diff
@@ -545,7 +580,22 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
       serviceAccountInfo.clusterServiceAccount, serviceAccountInfo.notebookServiceAccount, clusterRequest.jupyterExtensionUri, clusterRequest.jupyterUserScriptUri)
       .toJson.asJsObject.fields.mapValues(labelValue => labelValue.convertTo[String])
     // combine default and given labels
-    val allLabels = clusterRequest.labels ++ defaultLabels
+    val nbExtensions = clusterRequest.userJupyterExtensionConfig match {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Alternative one-liner:
```
val nbExtensions = clusterRequest.userJupyterExtensionConfig.map(_.nbExtensions).getOrElse(Map.empty)
```

File: jupyter-docker/jupyter_install_notebook_extension.sh [hunk 10]
```diff
@@ -1,10 +1,11 @@
...
-  JUPYTER_EXTENSION_NAME=`basename ${JUPYTER_EXTENSION_ARCHIVE%%.*}`
+  JUPYTER_EXTENSION=$1
+  JUPYTER_EXTENSION_NAME=`basename ${JUPYTER_EXTENSION%%.*}`
   mkdir ${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
-  tar -xzf ${JUPYTER_EXTENSION_ARCHIVE} -C${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
+  tar -xzf ${JUPYTER_EXTENSION} -C${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
   jupyter nbextension install ${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}/ --user
   jupyter nbextension enable ${JUPYTER_EXTENSION_NAME}/main
-fi 
+fi
```
Comment (kyuksel) [inline jupyter-docker/jupyter_install_notebook_extension.sh]: Nit: add new line

File: jupyter-docker/jupyter_install_server_extension.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (kyuksel) [inline jupyter-docker/jupyter_install_server_extension.sh]: I wonder if this is a good opportunity to organize the `jupyter-docker` folder a bit, perhaps by creating a `scripts` folder with subfolders like `extension-installation`, `pip-installation`, etc.

File: jupyter-docker/jupyter_install_combined_extension.sh [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+if [ -n "$1" ]; then
+  JUPYTER_EXTENSION=$1
+  JUPYTER_EXTENSION_NAME=`basename ${JUPYTER_EXTENSION%%.*}`
+  mkdir ${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
+  tar -xzf ${JUPYTER_EXTENSION} -C${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
+  pip install -e ${JUPYTER_HOME}/${JUPYTER_EXTENSION_NAME}
+  jupyter serverextension enable --py ${JUPYTER_EXTENSION_NAME}
+  jupyter nbextension install --py ${JUPYTER_EXTENSION_NAME}
+  jupyter nbextension enable --py ${JUPYTER_EXTENSION_NAME}
+fi
```
Comment (kyuksel) [inline jupyter-docker/jupyter_install_combined_extension.sh]: Nit: add new line

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterMonitoringSpec.scala [hunk 13]
```diff
@@ -239,9 +239,26 @@ class ClusterMonitoringSpec extends FreeSpec with LeonardoTestUtils with Paralle
           }
         }
       }
-
     }
 
+    "should install multiple user specified notebook extensions" in withWebDriver { implicit driver =>
+      val clusterName = ClusterName("user-jupyter-ext" + makeRandomId())
+      withCleanBillingProject(hermioneCreds) { projectName =>
```
Comment (kyuksel) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterMonitoringSpec.scala]: You can reduce some of this boiler plate (lines 246-249) by rebasing your branch and using Rob's `withProject` method, added with in [this PR](https://github.com/DataBiosphere/leonardo/pull/331/files#diff-7d71faa33c09b3f5553be71dd25769d5L43).

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterMonitoringSpec.scala [hunk 14]
```diff
@@ -239,9 +239,26 @@ class ClusterMonitoringSpec extends FreeSpec with LeonardoTestUtils with Paralle
...
+            nbExt.get should include("jupyter-gmaps/extension  enabled")
+            nbExt.get should include("pizzabutton/index  enabled")
+            nbExt.get should include("my_ext/main  enabled")
+            val serverExt = notebookPage.executeCell("! jupyter serverextension list")
+            serverExt.get should include("pizzabutton  enabled")
+            serverExt.get should include("jupyterlab  enabled")
+          }
+        }
+      }
+    }
```
Comment (kyuksel) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterMonitoringSpec.scala]: It may be good to add unit tests for some corner case scenarios such as permutations of one or more unspecified extensions as well as error cases such as invalid extension specifications.

File: src/main/resources/swagger/api-docs.yaml [hunk 15]
```diff
@@ -653,6 +653,9 @@ definitions:
       labels:
         type: object
         description: The labels to be placed on the cluster. Of type Map[String,String]
+      userJupyterExtensionConfig:
+        $ref: "#/definitions/UserJupyterExtensionConfig"
+        description: Jupyter extensions to be installed in the notebook
       jupyterExtensionUri:
```
[the remark this answers is not in the document]
Comment (vkumra-broad) [reply src/main/resources/swagger/api-docs.yaml]: Yep

File: src/main/resources/jupyter/init-actions.sh [hunk 16]
```diff
@@ -145,13 +147,53 @@ if [[ "${ROLE}" == 'Master' ]]; then
...
+    if [ ! -z ${JUPYTER_NB_EXTENSIONS} ] ; then
+      for ext in ${JUPYTER_NB_EXTENSIONS}
+      do
+        log 'Installing Jupyter NB extension...'
+        if [[ $ext == 'gs://'* && $ext == *'.tar.gz' ]]; then
+          gsutil cp $ext /etc
+          JUPYTER_EXTENSION_ARCHIVE=`basename $ext`
+          docker cp /etc/${JUPYTER_EXTENSION_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
+          docker exec -d ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/jupyter_install_notebook_extension.sh ${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
+        else
```
[the remark this answers is not in the document]
Comment (vkumra-broad) [reply src/main/resources/jupyter/init-actions.sh]: The extensions can also be python packages that we install using pip

File: src/main/resources/swagger/api-docs.yaml (hunk 6 above)
Comment (vkumra-broad) [reply src/main/resources/swagger/api-docs.yaml]: what about something like this:

'Optional, map of extension name and nbExtension. The nbExtension can either be a tar.gz file on google storage or a python package. Example, {"ext1":"gs:/bucket/extension.tar.gz", "pipExt":"python-packkage"}'

Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: Yes, that sounds good. 👍 

Similarly below for `serverExtensions`

File: src/main/resources/swagger/api-docs.yaml [hunk 17]
```diff
@@ -719,7 +722,19 @@ definitions:
...
+    description: 'Specification of Jupyter Extensions to be installed on the cluster'
+    properties:
+      nbExtensions:
+        type: object
+        description: |
+          Optional, map of extension name and notebook
+      serverExtensions:
+        type: object
+        description: Optional, map of extension name and server extension
+      combinedExtensions:
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: It sounds like `bundler` is yet another extension type? I think for this PR we can stick with `combined` and not implement `bundler`.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 18]
```diff
@@ -301,35 +301,38 @@ trait ClusterComponent extends LeoComponent {
...
     }
 
-    private def unmarshalClustersWithLabels(clusterLabels: Seq[(ClusterRecord, Option[LabelRecord])]): Seq[Cluster] = {
+    private def unmarshalClustersWithLabels(clusterLabels: Seq[(ClusterRecord, Option[LabelRecord], Option[ExtensionRecord])]): Seq[Cluster] = {
       // Call foldMap to aggregate a Seq[(ClusterRecord, LabelRecord)] returned by the query to a Map[ClusterRecord, Map[labelKey, labelValue]].
-      val clusterLabelMap: Map[ClusterRecord, LabelMap] = clusterLabels.toList.foldMap { case (clusterRecord, labelRecordOpt) =>
-        val labelMap = labelRecordOpt.map(labelRecordOpt => labelRecordOpt.key -> labelRecordOpt.value).toMap
-        Map(clusterRecord -> labelMap)
+      val clusterLabelMap: Map[ClusterRecord, (Map[String, List[String]], List[ExtensionRecord])] = clusterLabels.toList.foldMap { case (clusterRecord, labelRecordOpt, extensionOpt) =>
+        val labelMap = labelRecordOpt.map(labelRecordOpt => labelRecordOpt.key -> List(labelRecordOpt.value)).toMap
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: I'm curious why `labelMap` was changed from a `Map[String, String]` to a `Map[String, List[String]]`? So we can support multiple labels with the same name?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 19]
```diff
@@ -301,35 +301,38 @@ trait ClusterComponent extends LeoComponent {
...
+        val extList =  extensionOpt.toList
+        Map(clusterRecord -> (labelMap, extList))
       }
 
       // Unmarshal each (ClusterRecord, Map[labelKey, labelValue]) to a Cluster object
-      clusterLabelMap.map { case (clusterRec, labelMap) =>
-        unmarshalCluster(clusterRec, Seq.empty, List.empty, labelMap)
+      clusterLabelMap.map { case (clusterRec, (labelMap, extList)) =>
+        unmarshalCluster(clusterRec, Seq.empty, List.empty, labelMap.mapValues(_.toSet.head), extList)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Instead of just taking the first label, maybe we should concatenate them into a string:
```
labelMap.mapValues(_.toSet.mkString(", "))
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ExtensionComponent.scala [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+  import profile.api._
+
+  class ExtensionTable(tag: Tag) extends Table[ExtensionRecord](tag, "NOTEBOOK_EXTENSIONS") {
+    def clusterId = column[Long]("clusterId")
+
+    def extensionType = column[String]("extensionType", O.Length(254))
+
+    def name = column[String]("name", O.Length(254))
+
+    def path = column[String]("path", O.Length(254))
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ExtensionComponent.scala]: Since this is an arbitrary gs:// path, we might want to bump the length to 1024. The CLUSTER table does this as well: https://github.com/DataBiosphere/leonardo/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala#L69-L70

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ExtensionComponent.scala [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        case Some(ext) => extensionQuery ++= marshallExtensions(clusterId, ext)
+        case None => DBIO.successful(0)
+      }
+    }
+
+    def getAllForCluster(clusterId: Long): DBIO[UserJupyterExtensionConfig] = {
+      extensionQuery.filter {
+        _.clusterId === clusterId
+      }.result map { recs =>
+        val nbExtensions = (recs.filter(_.extensionType == "nbExtension") map { rec => rec.name -> rec.value }).toMap
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ExtensionComponent.scala]: Maybe make the extension types constants (or better yet, enums) in LeonardoModel instead of hardcoded strings here.

File: src/main/resources/swagger/api-docs.yaml (hunk 6 above)
Comment (kyuksel) [reply src/main/resources/swagger/api-docs.yaml]: I like it, particularly the example . Note the typo in pack**k**age in case you copy/paste :)

Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: And gs:/*/*

File: src/main/resources/swagger/api-docs.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (kyuksel) [reply src/main/resources/swagger/api-docs.yaml]: To clarify: I wasn't aware that `combined` and `bundler` extensions were distinct, and wasn't advocating for implementing both. For my edification, could you send me a link to info on `combined` extensions, perhaps like [this](http://jupyter-notebook.readthedocs.io/en/stable/extending/bundler_extensions.html) on the `bundler`?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala (hunk 19 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: Faces: we'll leave this as is

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 22]
```diff
@@ -421,30 +421,48 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
+  private[service] def validateBucketObjectUri(userEmail: WorkbenchEmail, googleProject: GoogleProject, gcsUri: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
+
+    val gcsUriOpt = parseGcsPath(gcsUri)
     gcsUriOpt match {
-      case None => Future.successful(())
-      case Some(gcsPath) =>
+      case Right(gcsPath) =>
         if (gcsPath.toUri.length > bucketPathMaxLength)
-          throw BucketObjectException(gcsPath)
+          throw BucketObjectException(gcsUri)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Since this returns a Future, should probably do `Future.failed(BucketObjectException(gcsUri))` instead of throwing exceptions directly.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 23]
```diff
@@ -421,30 +421,48 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
...
-              case false => throw BucketObjectException(gcsPath)
+              case false => throw BucketObjectException(gcsPath.toUri)
             } recover {
               case e: HttpResponseException if e.getStatusCode == StatusCodes.Forbidden.intValue =>
                 logger.error(s"User $userEmail does not have access to ${gcsPath.bucketName} / ${gcsPath.objectName}")
                 throw BucketObjectAccessException(userEmail, gcsPath)
             }
           case None => Future.successful(())
         }
+      case Left(err) => throw BucketObjectException(gcsUri)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: `Future.failed(BucketObjectException(gcsUri))`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 24]
```diff
@@ -421,30 +421,48 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     } getOrElse Future.successful(())
   }
 
-  private[service] def validateBucketObjectUri(userEmail: WorkbenchEmail, googleProject: GoogleProject, gcsUriOpt: Option[GcsPath])(implicit executionContext: ExecutionContext): Future[Unit] = {
 
+  private def validateClusterRequestBucketObjectUri(userEmail: WorkbenchEmail, googleProject: GoogleProject, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    for {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: We should probably also validate the deprecated `jupyterExtensionUri`, since it's still being used below.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala [hunk 25]
```diff
@@ -123,6 +124,9 @@ class AuthProviderSpec extends FreeSpec with ScalatestRouteTest with Matchers wi
       // get status
       val clusterStatus = leo.getActiveClusterDetails(userInfo, project, name1).futureValue
 
+      println(cluster1)
```
Comment (rtitle) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala]: Remove printlns

**Comment by rtitle** [comment]
Looks good, just a few more minor comments above. 

Re unit tests: it looks like the DB stuff and labels are covered. I'd like a case that validates the new templated fields in `init-actions.sh`. That should be sufficient for unit testing.

For integration tests: it looks like `withNewErroredCluster` is no longer working. IIRC it used to generate an errored cluster by specifying an extension URI which doesn't exist. Something in the new code is not causing that to fail anymore -- would be good to understand why.

File: src/main/resources/swagger/api-docs.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (vkumra-broad) [reply src/main/resources/swagger/api-docs.yaml]: http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Distributing%20Jupyter%20Extensions%20as%20Python%20Packages.html#Install-and-enable-the-server-extension-and-nbextension

This is what I was trying to make easier with the combined extensions.

[the remark this answers is not in the document]
Comment (kyuksel) [reply src/main/resources/swagger/api-docs.yaml]: I see. Thanks!

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

File: src/main/resources/swagger/api-docs.yaml [hunk 26]
```diff
@@ -738,7 +741,22 @@ definitions:
...
+    description: 'Specification of Jupyter Extensions to be installed on the cluster'
+    properties:
+      nbExtensions:
+        type: object
+        description: |
+          Optional, map of extension name and nbExtension. The nbExtension can either be a tar.gz file on google storage or a python package.
+          Example, {"ext1":"gs://bucket/extension.tar.gz", "pipExt":"python-package"}
+      serverExtensions:
+        type: object
+        description: Optional, map of extension name and server extension. The serverExtensions can either be a tar.gz file on google storage or a python package.
```
Comment (rtitle) [inline src/main/resources/swagger/api-docs.yaml]: Does this multiline format work in Swagger? Or do you have to do
```
  description: |
    Description text with
    multiple lines
```
as above?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 24 above)
Comment (vkumra-broad) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: We are adding the extension to nbExtensions map and validating it with the rest of the extensions

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest

**Comment by vkumra-broad** [comment]
jenkins retest
